### PR TITLE
added single connection for all trade pairs

### PR DESCRIPTION
**Pull request by NikGorbunov**

### CODE REVIEW

File: alarm/binance_utils.py [hunk 1]
```diff
@@ -26,14 +26,11 @@ def close_binance_sockets(sockets):
...
     websocket.enableTrace(False)
-    socket_urls = [f'wss://stream.binance.com:9443/ws/{currency}@kline_{interval}' for currency in trade_pairs]
-    sockets = []
-    for socket_url in socket_urls:
-        socket = websocket.create_connection(socket_url, sslopt={'cert_reqs': ssl.CERT_NONE})
-        sockets.append(socket)
-
-    logger.info(f"Sockets connected: {', '.join(socket_urls)}.")
-    return sockets
+    symbols = '/'.join([f'{trade_pair}@kline_{interval}' for trade_pair in trade_pairs])
```
Comment (oleksandr-shvab) [inline alarm/binance_utils.py:29]: symbols are not the best naming, try to describe what contains in this string

File: alarm/binance_utils.py [hunk 2]
```diff
@@ -26,14 +26,11 @@ def close_binance_sockets(sockets):
...
+    socket_url = f'wss://stream.binance.com:9443/ws/{symbols}'
+    socket = websocket.create_connection(socket_url, sslopt={'cert_reqs': ssl.CERT_NONE})
+    logger.info(f"Sockets connected: {socket_url}).")
+    return [socket]
```
Comment (oleksandr-shvab) [inline alarm/binance_utils.py:33]: if we have one socket for all coins, do we need to put it in the list?

Comment (NikGorbunov) [reply alarm/binance_utils.py:33]: Yes, because method resv() does not accept the string.

File: alarm/binance_utils.py (hunk 1 above)
Comment (NikGorbunov) [reply alarm/binance_utils.py:29]: You're right, already changed the naming

File: alarm/binance_utils.py (hunk 2 above)
Comment (oleksandr-shvab) [reply alarm/binance_utils.py:34]: Ok

File: alarm/binance_utils.py [hunk 3]
```diff
@@ -26,14 +26,12 @@ def close_binance_sockets(sockets):
...
-    socket_urls = [f'wss://stream.binance.com:9443/ws/{currency}@kline_{interval}' for currency in trade_pairs]
-    sockets = []
-    for socket_url in socket_urls:
-        socket = websocket.create_connection(socket_url, sslopt={'cert_reqs': ssl.CERT_NONE})
-        sockets.append(socket)
-
-    logger.info(f"Sockets connected: {', '.join(socket_urls)}.")
-    return sockets
+    trade_pairs_and_intervals_for_socket_url = '/'.join(
+        [f'{trade_pair}@kline_{interval}' for trade_pair in trade_pairs])
```
Comment (oleksandr-shvab) [inline alarm/binance_utils.py:30]: In Alex's code we had "for currency in trade_pairs", why we changed to "for trade_pair in trade_pairs"?

File: alarm/binance_utils.py [hunk 4]
```diff
@@ -44,4 +43,4 @@ def parse_candle_from_websocket_update(data):
     trade_pair = candle['s']
     trade_pair_lower_case = trade_pair.lower()
 
-    return candle_high_price, candle_low_price, trade_pair_lower_case
+    return candle_high_price, candle_low_price, trade_pair_lower_case
```
Comment (shuryhin-oleksandr) [inline alarm/binance_utils.py:46]: Pay attention to this. PEP-8 requires an empty line in the end of the file.

File: alarm/models.py [hunk 5]
```diff
@@ -55,7 +55,7 @@ def penultimate_for_trade_pair(cls, trade_pair):
 
     @classmethod
     @transaction.atomic
-    def save_as_recent(cls, trade_pair, high_price, low_price):
+    def update_or_create_prices(cls, trade_pair, high_price, low_price):
```
Comment (shuryhin-oleksandr) [inline alarm/models.py:58]: Bad naming, as method does not operate on prices, it operates on candles.
What's worse, it does not update anything, it creates and deletes.
You can remember of this case such as an anti-pattern.

File: alarm/utils.py [hunk 6]
```diff
@@ -20,13 +20,14 @@ def any_of_trade_pair_thresholds_is_broken(trade_pair):
...
-        threshold_broken = threshold_is_broken(threshold.price, last_candle, penultimate_candle)
-        logger.info(f"{str(trade_pair).upper()}; "
-                    f"candles: {penultimate_candle}, {last_candle}; "
-                    f"threshold: {threshold}; "
-                    f"threshold broken: {threshold_broken}")
-        if threshold_broken:
-            return True
-
-    return False
+    if last_candle is not None and penultimate_candle is not None:
```
Comment (shuryhin-oleksandr) [inline alarm/utils.py:23]: if last_candle is None or penultimate_candle is None:
   return False
   
Flat is better than nested. Decrease a nesting level.

Comment (NikGorbunov) [reply alarm/utils.py]: ++

File: alarm/models.py (hunk 5 above)
Comment (NikGorbunov) [reply alarm/models.py]: ++

File: alarm/binance_utils.py (hunk 4 above)
Comment (NikGorbunov) [reply alarm/binance_utils.py]: ++

File: alarm/binance_utils.py [hunk 7]
```diff
@@ -18,22 +18,21 @@ def get_binance_trade_pairs():
...
-        sockets.append(socket)
-
-    logger.info(f"Sockets connected: {', '.join(socket_urls)}.")
-    return sockets
+    trade_pairs_and_intervals_for_socket_url = '/'.join(
+        [f'{trade_pair}@kline_{interval}' for trade_pair in trade_pairs])
+
+    socket_url = f'wss://stream.binance.com:9443/ws/{trade_pairs_and_intervals_for_socket_url}'
+    socket = websocket.create_connection(socket_url, sslopt={'cert_reqs': ssl.CERT_NONE})
+    logger.info(f"Sockets connected: {socket_url}.")
```
Comment (oleksandr-shvab) [inline alarm/binance_utils.py:33]: We have one socket but here "Sockets"

Comment (NikGorbunov) [reply alarm/binance_utils.py:33]: ++

**Review by shuryhin-oleksandr — APPROVED** [review]

**Pull request merged**